core/utils/utils.go

```diff
@@ -418,46 +418,48 @@ func WaitGroupChan(wg *sync.WaitGroup) <-chan struct{} {
 // channel is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (ctx context.Context, cancel context.CancelFunc) {
-	ctx, cancel = context.WithCancel(parentCtx)
-
-	go func() {
-		select {
-		case <-chStop:
-		case <-ctx.Done():
-		}
-		cancel()
-	}()
-
-	return ctx, cancel
+func WithCloseChan(parentCtx context.Context, chStop <-chan struct{}) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithCancel(parentCtx))
 }
 
 // ContextFromChan creates a context that finishes when the provided channel
 // receives or is closed.
-// When channel closes, the ctx.Err() will always be context.Canceled
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
 func ContextFromChan(chStop <-chan struct{}) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithCancel(context.Background())
-	go func() {
-		select {
-		case <-chStop:
-			cancel()
-		case <-ctx.Done():
-		}
-	}()
-	return ctx, cancel
+	return StopChan(chStop).NewCtx()
 }
 
-// ContextFromChanWithDeadline creates a context with a deadline that finishes when the provided channel
+// ContextFromChanWithTimeout creates a context with a timeout that finishes when the provided channel
 // receives or is closed.
 // NOTE: Spins up a goroutine that exits on cancellation.
 // REMEMBER TO CALL CANCEL OTHERWISE IT CAN LEAD TO MEMORY LEAKS
-func ContextFromChanWithDeadline(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
-	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+func ContextFromChanWithTimeout(chStop <-chan struct{}, timeout time.Duration) (context.Context, context.CancelFunc) {
+	return StopChan(chStop).CancelCtx(context.WithTimeout(context.Background(), timeout))
+}
+
+type StopChan <-chan struct{}
+
+func NewStopChan() (ch StopChan, stop func()) {
+	c := make(chan struct{})
+	return c, func() { close(c) }
+}
+
+// NewCtx returns a new background [context.Context] that will be cancelled when the StopChan is closed.
```

Comment:
StopChan is closed -> stop function is called

Reply:
Not necessarily. If you wrap an existing `stopCh chan struct{}`, then the existing direct call of `close(stopCh)` will continue to work. The constructor stop func is actually just a quirk that is only necessary b/c the type is a `<-chan`. It had to be that way to drop in without disrupting too much else, but I'm still fiddling so it might simplify away.

Reply:
that's true, but that's not exposed yet.

if we want to add this abstraction, then we should take ownership of the underlying channel so that there aren't multiple entry points to do the same thing. that is, we shouldn't enable type casting from a vanilla ch to a StopChan and allow both close(vanilla chan) and stop() to do the same thing

Reply:
It is exposed, by design. Why shouldn't we do that? It's a simpler, composable, more efficient abstraction. The original version didn't require a constructor at all, b/c you can just `make(StopChan)`. The only reason the constructor (and the func()) exist is to support the need to use `<-chan` with the existing methods. I would like to get back to that version if it's not too disruptive, because this constructor variant does not drop right in - it would be a much more intrusive change.